### ISSUE TITLE
Strange failure of test_signalhandler.ml when running with the debug runtime

### ISSUE THREAD

**Issue by abbysmal** [ocaml/ocaml#10997]
I was investigating a failure that came up when making adjustments to the debug runtime testsuite run on the CI.

I stumbled upon this headscratcher in the process and I am having a hard time wrapping my head around the issue so I am posting this here while I investigate a bit more.

The gist of it is that this testcase fails on trunk under these specific parameters:
`OCAMLRUNPARAM=v=0 USE_RUNTIME=d make one TEST=tests/callback/test_signalhandler.ml`

Now the interesting thing is that it does not trigger an abort, rather, it seems to segfault here:
```
Program received signal SIGSEGV, Segmentation fault.
camlStdlib__output_string_248 () at stdlib.ml:367
367	  unsafe_output_string oc s 0 (string_length s)
(rr) bt
#0  camlStdlib__output_string_248 () at stdlib.ml:367
#1  0x00000000004372e8 in camlTest_signalhandler__entry ()
#2  0x0000000000434260 in caml_program ()
#3  <signal handler called>
#4  caml_startup (argv=<optimized out>) at startup_nat.c:136
#5  caml_main (argv=<optimized out>) at startup_nat.c:142
#6  0x0000000000433f2c in main (argc=<optimized out>, argv=<optimized out>) at main.c:37
```

Trying to backtrack a bit, my early investigations seemed to suggest there may be a very very strange play between `caml_gc_log`/`caml_gc_message` and the sighandler dispatch, causing some kind of corruption.

The why `caml_gc_log` specifically, is because if I backtrack to the faulty memory access, it seems that the this area is last written by a `local_memcpy` call within `pthread_sigmask`, as called in `caml_execute_signal`.
Before these writes, `caml_gc_log` seems to be the last few calls writing to this address.

This may be a red herring, though it would make sense because this only happening within the debug runtime is easily linked to these GC debug messages, only happening in said runtime.

For reference, here is a simplified testcase, which is a shaven `callback/test_signalhandler.ml` still exhibiting this behavior.
```ocaml
let rec tak (x, y, z as _tuple) =                                                                                                                                         
  if x > y then tak(tak (x-1, y, z), tak (y-1, z, x), tak (z-1, x, y))                                                                                                    
           else z                                                                                                                                                         
                                                                                                                                                                          
(* Test callbacks performed to handle signals *)                                                                                                                          
                                                                                                                                                                          
let sighandler signo =                                                                                                                                                    
(*                                                                                                                                                                        
  print_string "Got signal, triggering garbage collection...";                                                                                                            
  print_newline();                                                                                                                                                        
*)                                                                                                                                                                        
  (* Thoroughly wipe the minor heap *)                                                                                                                                    
  ignore (tak (18, 12, 6))                                                                                                                                                
                                                                                                                                                                          
external unix_getpid : unit -> int = "unix_getpid" [@@noalloc]                                                                                                            
external unix_kill : int -> int -> unit = "unix_kill" [@@noalloc]                                                                                                         
                                                                                                                                                                          
let callbacksig () =                                                                                                                                                      
  let pid = unix_getpid() in                                                                                                                                              
  (* Allocate a block in the minor heap *)                                                                                                                                
  let s = String.make 5 'b' in                                                                                                                                            
  (* Send a signal to self.  We want s to remain in a register and                                                                                                        
     not be spilled on the stack, hence we declare unix_kill                                                                                                              
     [@@noalloc]. *)                                                                                                                                                      
  unix_kill pid Sys.sigusr1;                                                                                                                                              
  (* Allocate some more so that the signal will be tested *)                                                                                                              
  let u = (s, s) in                                                                                                                                                       
  fst u                                                                                                                                                                   
                                                                                                                                                                          
let _ =                                                                                                                                                                   
  Sys.set_signal Sys.sigusr1 (Sys.Signal_handle sighandler);                                                                                                              
  print_string(callbacksig ()); print_newline()      
```
To be compiled with `./ocamlopt.opt -I runtime/ -nostdlib -I stdlib/ -I otherlibs/unix/ unix.cmxa -g -runtime-variant=d -o a.out test.ml` and run with `OCAMLRUNPARAM=v=0 ./a.out`
(although the stacktrace is completely different on this one than `test_signalhandler`.)

One more thing but I did not look into it yet, it also appear the if I comment out signal processing in `unix_kill` `test_signalhandler.ml` and the provided testcase stops segfaulting. (but the signal handler is still dispatched later.)

This last comment is not a "fix" by any mean, but maybe there's a subtlety I'm missing here.

See: https://github.com/ocaml/ocaml/blob/trunk/otherlibs/unix/kill.c#L30

**Comment by xavierleroy** [comment]
This comment is a smoking gun:
```
  (* Send a signal to self.  We want s to remain in a register and
     not be spilled on the stack, hence we declare unix_kill
     [@@noalloc]. *)
```
If `unix_kill` processes signals, it can call back into OCaml (like here), so it needs to go through `caml_c_call` so that everything is properly registered with the GC.  So, declaring `unix_kill` as `@@noalloc` is wrong.

**Comment by xavierleroy** [comment]
How does it go with the fix at #11033 ?
